domain/machine/state/state.go

```diff
@@ -187,10 +187,42 @@ func (st *State) GetMachineLife(ctx context.Context, machineName machine.Name) (
 	})
 
 	if err != nil && errors.Is(err, sqlair.ErrNoRows) {
-		return nil, errors.NotFoundf("machine %q", machineName)
+		return nil, errors.NotFoundf("machine %q", mName)
+	}
+
+	return &lifeResult, errors.Annotatef(err, "getting life status for machines %q", mName)
+}
+
+// IsController returns whether the machine is a controller machine.
+// It returns a NotFound if the given machine doesn't exist.
+func (st *State) IsController(ctx context.Context, mName machine.Name) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	machineNameParam := machineName{Name: mName}
+	query := `SELECT is_controller AS &machineIsController.is_controller FROM machine WHERE name = $machineName.name`
+	queryStmt, err := st.Prepare(query, machineNameParam, machineIsController{})
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	var isController bool
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		result := machineIsController{}
```

Comment:
This whole function is simple enough to declare this at the top, pass it to prepare, and return its inner bool without reassignment.